src/serialize/json/denormalize.ts

```diff
@@ -0,0 +1,317 @@
+limitations under the License.
+
+SPDX-License-Identifier: Apache-2.0
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+import { PackageURL } from 'packageurl-js'
+import { URL } from 'url'
+
+import * as Models from '../../models'
+import type { Protocol as Spec, Version as SpecVersion } from '../../spec'
+import { SpecVersionDict, UnsupportedFormatError } from '../../spec'
+import type { DenormalizerOptions } from '../types'
+import type { Normalized } from './types'
+
+export class Factory {
+  #spec?: Spec
+
+  get spec (): Spec | undefined {
+    return this.#spec
+  }
+
+  set spec (spec: Spec | undefined) {
+    this.#spec = spec
+  }
+
+  makeForBom (): BomDenormalizer {
+    return new BomDenormalizer(this)
+  }
+
+  makeForMetadata (): MetadataDenormalizer {
+    return new MetadataDenormalizer(this)
+  }
+
+  makeForComponent (): ComponentDenormalizer {
+    return new ComponentDenormalizer(this)
+  }
+
+  makeForTool (): ToolDenormalizer {
+    return new ToolDenormalizer(this)
+  }
+
+  makeForOrganizationalContact (): OrganizationalContactDenormalizer {
+    return new OrganizationalContactDenormalizer(this)
+  }
+
+  makeForOrganizationalEntity (): OrganizationalEntityDenormalizer {
+    return new OrganizationalEntityDenormalizer(this)
+  }
+
+  makeForHash (): HashDenormalizer {
+    return new HashDenormalizer(this)
+  }
+
+  makeForLicense (): LicenseDenormalizer {
+    return new LicenseDenormalizer(this)
+  }
+
+  makeForSWID (): SWIDDenormalizer {
+    return new SWIDDenormalizer(this)
+  }
+
+  makeForExternalReference (): ExternalReferenceDenormalizer {
+    return new ExternalReferenceDenormalizer(this)
+  }
+
+  makeForAttachment (): AttachmentDenormalizer {
+    return new AttachmentDenormalizer(this)
+  }
+
+  makeForProperty (): PropertyDenormalizer {
+    return new PropertyDenormalizer(this)
+  }
+
+  makeForUrl (): (url: string, options: DenormalizerOptions) => (URL | string) {
+    return (u, o) => denormalizeUrl(u, o, this)
+  }
+}
+
+interface JsonDenormalizer<TModel, TNormalized> {
+  denormalize: (data: TNormalized, options: DenormalizerOptions) => TModel | undefined
+}
+
+abstract class BaseJsonDenormalizer<TModel, TNormalized = object> implements JsonDenormalizer<TModel, TNormalized> {
+  protected readonly _factory: Factory
+
+  constructor (factory: Factory) {
+    this._factory = factory
+  }
+
+  get factory (): Factory {
+    return this._factory
+  }
+
+  abstract denormalize (data: TNormalized, options: DenormalizerOptions): TModel | undefined
+}
+
+export class BomDenormalizer extends BaseJsonDenormalizer<Models.Bom> {
+  denormalize (data: Normalized.Bom, options: DenormalizerOptions): Models.Bom {
+    const spec = SpecVersionDict[data.specVersion as SpecVersion]
+    if (spec === undefined) {
+      throw new UnsupportedFormatError(`Spec version ${data.specVersion} is not supported.`)
+    }
+    this.factory.spec = spec
```

Comment:
https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506/files#diff-facac0ec4128527fecf05b9814c75d81c08dfe23a5bc7c9f26611f6e224b3c51R112-R117

this should be part of the tool that does initial analysis.
reading which spec version is applied should be done before the factory is initialized.

Reply:
you might wonder how the factory should be injectable, while it is not fully initialized? 
possible solution: Inject a factory method, or the class type, instead of an object o the class.